workshops/models.py

```diff
@@ -1950,3 +1993,152 @@ class TrainingRequest(ActiveMixin, CreatedUpdatedMixin, models.Model):
+            raise ValidationError({'person': 'Accepted training request must '
+                                             'be matched to a person.'})
+
+    def get_absolute_url(self):
+        return reverse('trainingrequest_details', args=[self.pk])
+
+
+class TrainingRequirement(models.Model):
+    name = models.CharField(max_length=STR_MED)
+
+    # Determines whether TrainingProgress.url is required (True) or must be
+    # null (False).
+    url_required = models.BooleanField(default=False)
+
+    # Determines whether TrainingProgress.event is required (True) or must be
+    # null (False).
+    event_required = models.BooleanField(default=False)
+
+    def __str__(self):
+        return self.name
+
+    class Meta:
+        ordering = ['name']
+
+
+@reversion.register
+class TrainingProgress(CreatedUpdatedMixin, models.Model):
+    trainee = models.ForeignKey(Person, on_delete=models.PROTECT)
+
+    # Mentor/examiner who evaluates homework / session. May be null when a
+    # trainee submits their homework.
+
+    # FIXME Maybe let's rename it as evaluated_by?
+    awarded_by = models.ForeignKey(Person,
+                                   on_delete=models.PROTECT,
+                                   null=True, blank=True,
+                                   related_name='+')
+    requirement = models.ForeignKey(TrainingRequirement,
+                                    on_delete=models.PROTECT)
+
+    STATES = (
+        ('n', 'Not-evaluated yet'),
+        ('f', 'Failed'),
+        ('p', 'Passed'),
+    )
+    state = models.CharField(choices=STATES, default='p', max_length=1)
+
+    # When we end training and trainee has gone silent, or passed their
+    # deadline, we set this field to True.
+    discarded = models.BooleanField(
+        default=False,
+        verbose_name='Discarded -- check if the trainee has gone silent '
+                     'or passed their training deadline.')
+
+    event = models.ForeignKey(Event, null=True, blank=True)
+    url = models.URLField(null=True, blank=True)
+    notes = models.TextField(blank=True)
+
+    def get_absolute_url(self):
+        return reverse('trainingprogress_edit', args=[str(self.id)])
+
+    def clean(self):
+        if self.requirement.url_required and not self.url:
+            msg = 'In the case of {}, this field is required.'.format(self.requirement)
+            raise ValidationError({'url': msg})
+        elif not self.requirement.url_required and self.url:
+            msg = 'In the case of {}, this field must be left empty.'.format(self.requirement)
+            raise ValidationError({'url': msg})
+
+        if self.requirement.event_required and not self.event:
+            msg = 'In the case of {}, this field is required.'.format(self.requirement)
+            raise ValidationError({'event': msg})
+        elif not self.requirement.event_required and self.event:
+            msg = 'In the case of {}, this field must be left empty.'.format(self.requirement)
```

Comment:
This validation requires that `requirement` is already cleaned, so it cannot go to `clean_FIELD` methods.